### PR TITLE
Create unit tests

### PR DESCRIPTION
**Pull request by ihhyy**

### CODE REVIEW

**Review by alesyalis — APPROVED** [review]

**Review by Xodoc — APPROVED** [review]

File: BL/Services/Validator.cs [hunk 1]
```diff
@@ -14,7 +14,7 @@ public void ValidateInput(string input)
 
         public void ValidateOutput(Weather output)
         {
-            if (output.Cod == 404)
+            if (output == null)
                 throw new Exception("Incorrect city name or city not found");
```
Comment (denja2) [inline BL/Services/Validator.cs]: Don't use generic exception

File: Tests/Services/ValidatorTest.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
+﻿using BL.Interfaces;
+using BL.Services;
+using DAL.Entities;
+using System;
+using Tests.Fixtures;
+using Xunit;
+
+namespace Tests.Services
+{
+    public class ValidatorTest
```
Comment (denja2) [inline Tests/Services/ValidatorTest.cs]: ValidatorServiceTest

File: Tests/Services/ValidatorTest.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        private readonly WeatherFixture _weatherFixture;
+        private readonly IValidator _validator;
+
+        public ValidatorTest()
+        {
+            _weatherFixture = new WeatherFixture();
+            _validator = new ValidatorServices();
```
Comment (denja2) [inline Tests/Services/ValidatorTest.cs]: Why services?

File: Tests/Services/ValidatorTest.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        [Fact]
+        public void ValidateInput_IncorrectInput()
```
Comment (denja2) [inline Tests/Services/ValidatorTest.cs]: What about the naming

File: Tests/Services/ValidatorTest.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            //Arrange
+            var input = "";
```
Comment (denja2) [inline Tests/Services/ValidatorTest.cs]: String empty

File: Tests/Services/ValidatorTest.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            //Arrange
+            var output = _weatherFixture.GetWeather()[0];
+
+            //Act
+            Action result = () => _validator.ValidateOutput(output);
+            result.Invoke();
+
+            //Assert
+            Assert.True(true);
```
Comment (denja2) [inline Tests/Services/ValidatorTest.cs]: It's fun

File: BL/Interfaces/IWeatherService.cs [hunk 7]
```diff
@@ -1,9 +1,10 @@
-﻿using System.Threading.Tasks;
+﻿using DAL.Entities;
+using System.Threading.Tasks;
 
 namespace BL.Interfaces
 {
     public interface IWeatherService
     {
-        Task<string> GetWeatherByCytyNameAsync(string cityName, string key);
+        Task<Weather> GetWeatherByCytyNameAsync(string cityName, string key);
```
Comment (denja2) [inline BL/Interfaces/IWeatherService.cs]: string

File: Tests/Services/ValidatorTest.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        public void ValidateInput_IncorrectInput()
+        {
+            //Arrange
+            var input = "";
+
+            //Act
+            Action result = () => _validator.ValidateInput(input);
+
+            //Assert
+            Assert.Throws<Exception>(result);
```
Comment (denja2) [inline Tests/Services/ValidatorTest.cs]: Also check error message

**Review by denja2 — CHANGES_REQUESTED** [review]

File: BL/Interfaces/IWeatherService.cs (hunk 7 above)
Comment (ihhyy) [reply BL/Interfaces/IWeatherService.cs]: Done

File: BL/Services/Validator.cs (hunk 1 above)
Comment (ihhyy) [reply BL/Services/Validator.cs]: Done

File: Tests/Services/ValidatorTest.cs (hunk 2 above)
Comment (ihhyy) [reply Tests/Services/ValidatorTest.cs]: Done

File: Tests/Services/ValidatorTest.cs (hunk 3 above)
Comment (ihhyy) [reply Tests/Services/ValidatorTest.cs]: Resolved

File: Tests/Services/ValidatorTest.cs (hunk 5 above)
Comment (ihhyy) [reply Tests/Services/ValidatorTest.cs]: Done

File: Tests/Services/ValidatorTest.cs (hunk 6 above)
Comment (ihhyy) [reply Tests/Services/ValidatorTest.cs]: Remade

File: Tests/Services/ValidatorTest.cs (hunk 8 above)
Comment (ihhyy) [reply Tests/Services/ValidatorTest.cs]: Done

File: Tests/Services/ValidatorTest.cs (hunk 4 above)
Comment (ihhyy) [reply Tests/Services/ValidatorTest.cs]: Fixed

File: Tests/Services/ValidatorTest.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    public class ValidatorTest
+    {
+        private readonly IValidator _validator;
+
+        public ValidatorTest()
+        {
+            _validator = new Validator();
+        }
+
+        [Fact]
```
Comment (denja2) [inline Tests/Services/ValidatorTest.cs]: Theory check for nul

File: Tests/Services/WeatherServiceTest.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace Tests.Services
+{
+    public class WeatherServiceTest
+    {
+        private readonly WeatherFixture _weatherFixture;
+        private readonly IWeatherService _weatherService;
+        private readonly Mock<IValidator> _repoValidator;
+        private readonly Mock<IWeatherRepository> _repoMock;
+        private readonly static string _key = ConfigurationManager.AppSettings["APIKey"];
```
Comment (denja2) [inline Tests/Services/WeatherServiceTest.cs]: ??

File: DAL/Enums/ServerErrorStatuses.cs [hunk 11]
```diff
@@ -0,0 +1,18 @@
+﻿
+namespace DAL.Enums
+{
+    public enum ServerErrorStatuses
```
Comment (denja2) [inline DAL/Enums/ServerErrorStatuses.cs]: use built in

File: DAL/Repositories/WeatherRepository.cs [hunk 12]
```diff
@@ -12,21 +13,23 @@ public class WeatherRepository : IWeatherRepository
...
+        public async Task<Weather> GetWeatherByCityNameAsync(string cityName)
         {
-            var response = await _client.GetAsync($"{_API}q={cityName}&appid={key}&units=metric");
+            var response = await _client.GetAsync($"{_API}q={cityName}&appid={_key}&units=metric");
             var responseBody = await response.Content.ReadAsStringAsync();
             var weather = JsonConvert.DeserializeObject<Weather>(responseBody);
 
-            if (weather.Cod >= 500)
-                throw new Exception("Server error");
+            if (weather.Code >= 500)
```
Comment (denja2) [inline DAL/Repositories/WeatherRepository.cs]: ?

File: Tests/Services/WeatherServiceTest.cs [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            //Arrange
+            var expected = _weatherFixture.GetWeather().Where(w => w.Name == cityName).FirstOrDefault();
+            _repoMock.Setup(x => x.GetWeatherByCityNameAsync(It.IsAny<string>())).ReturnsAsync(expected);
+
+            //Act
+            var weather = await _weatherService.GetWeatherByCytyNameAsync(cityName);
+
+            //Assert
+            Assert.Equal(weather.Message, message);
```
Comment (denja2) [inline Tests/Services/WeatherServiceTest.cs]: check Badrequest is false

File: Tests/Services/WeatherServiceTest.cs [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        public async void GetWeatherAsync_InCorrectInput_ReturnMessageWithError(string cityName)
+        {
+            var expected = _weatherFixture.GetWeather().Where(w => w.Name == cityName).FirstOrDefault();
+            _repoMock.Setup(x => x.GetWeatherByCityNameAsync(It.IsAny<string>())).ReturnsAsync(expected);
+
+            //Act
+            var weather = await _weatherService.GetWeatherByCytyNameAsync(cityName);
+
+            //Assert
+            Assert.Equal("City not found or input was incorrect", weather.Message);
```
Comment (denja2) [inline Tests/Services/WeatherServiceTest.cs]: check flag and cases with errors

File: Tests/Services/WeatherServiceTest.cs (hunk 10 above)
Comment (ihhyy) [reply Tests/Services/WeatherServiceTest.cs]: Removed

File: Tests/Services/ValidatorTest.cs (hunk 9 above)
Comment (ihhyy) [reply Tests/Services/ValidatorTest.cs]: Done

File: Tests/Services/WeatherServiceTest.cs (hunk 13 above)
Comment (ihhyy) [reply Tests/Services/WeatherServiceTest.cs]: Done

File: Tests/Services/WeatherServiceTest.cs (hunk 14 above)
Comment (ihhyy) [reply Tests/Services/WeatherServiceTest.cs]: Done

**Review by alesyalis — APPROVED** [review]

**Pull request merged**